feat: add pre-commit hook support

ccoVeille opened the pull request:
Description
-------------

Add pre-commit.com support in mockery.

- Fixes #791

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

Version of Go used when building/testing:
---------------------------------------------

- [ ] 1.11
- [ ] 1.12
- [ ] 1.13
- [ ] 1.14
- [ ] 1.15
- [ ] 1.16
- [ ] 1.17
- [ ] 1.18
- [ ] 1.19
- [ ] 1.20
- [X] Not relevant

How Has This Been Tested?
---------------------------

Locally, by following the guide I provide in installation.md file

Checklist
-----------

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

liveFreeOrCode commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,9 @@
+  name: generate mocks
+  description: Generate mocks with mockery
+  language: golang
+  additional_dependencies:
+    - github.com/vektra/mockery/v2@latest
+  entry: mockery
```
Rather than using additional_dependencies, the pre-commit-hook will already have access to the file system.  We should just be able to use `go run .` to enact the `mockery ...` command I believe.

Also, we may only want this hook to run when `.go` files are in the commit.  What do you think?

```suggestion
  entry: go run .
  files: \.go$
```

ccoVeille replied:
I'm unsure it will launch the code you expect to run.

I will make some tests.

But, I think keeping mockery call here would be clearer.

liveFreeOrCode replied:
Using additional dependencies like this won't allow people to specify version in pre commit file because no matter what version they reference, you're pulling latest

ccoVeille replied:
Indeed, I'm unsure if there is something that can be done here, I will have to check

ccoVeille replied:
For now,it seems to me, that it would imply to maintain the version in sync in this file vs the tag. Quite a pain

sx-klaviyo replied:
Is it possible let the users assume the responsibility for installing mockery? If that's the case this should suffice
```yaml
---
- id: generate-mocks
  name: generate mocks
  description: Generate mocks with mockery
  language: golang
  entry: mockery
  pass_filenames: false
```

ccoVeille replied:
Yes, of course, it makes sense.

But then having 2 rules could be a something to consider

ccoVeille replied:
I think your approach is the most commonly used in fact

https://github.com/search?q=path%3A%22%2F.pre-commit-hooks.yaml%22+golang&type=code

liveFreeOrCode commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+repos:
+  - repo: https://github.com/ccoVeille/mockery
```
Is this just in here for testing right now?  Unless mockery uses itself to generate mocks, and wants to use the pre-commit functionality, I don't think this is required to get this to work for consumers.

ccoVeille replied:
Yes, only for the test. I forgot to mention about it in the PR

I had to create a fake "2.43.3" tag to be able to make it work.

liveFreeOrCode commented on docs/installation.md:
````diff
@@ -28,6 +28,32 @@ Generate all the mocks for your project:
+
+```yaml
+repos:
+  - repo: https://github.com/ccoVeille/mockery
````
```suggestion
  - repo: https://github.com/vektra/mockery
```

ccoVeille replied:
I kept it in sync to help mockery mainteners to be able to test

But yes, I will have to change this one

liveFreeOrCode commented on docs/installation.md:
````diff
@@ -28,6 +28,32 @@ Generate all the mocks for your project:
+```yaml
+repos:
+  - repo: https://github.com/ccoVeille/mockery
+    rev: v2
````
nit/nb: It would be great if this could be automatically updated on new releases.

```suggestion
    rev: v2.43.2
```

ccoVeille replied:
There is no need usually. precommit manage this well with this kind of tag

liveFreeOrCode commented on docs/installation.md:
````diff
@@ -28,6 +28,32 @@ Generate all the mocks for your project:
+Then install the hook with the following command:
+
+```shell
+pre-commit autoupdate
````
Ah, this is why you have it as `v2` above, calling for autoupdate will which bump the version to latest.  May be better to be explicit about the versioning system, and call for `pre-commit install` versus autoupdate.

ccoVeille replied:
Indeed, I'm used to use autoupdate for everything.

I think "pre-commit install" would also bump it locally. I will make some tests.

Then I will update the documentation based on my tests

ccoVeille commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,6 @@
+---
```
Note for myself

I will have to add Co-Authored-By metadata

ccoVeille commented:
anyway @liveFreeOrCode could you confirm it "works" as this ?

ccoVeille commented:
@liveFreeOrCode (or anyone) could you confirm ? thanks

Then I will remove the oddities present to make it works with my branch, and switch to vektra/mokery one

ccoVeille commented:
Hi @liveFreeOrCode 

You commented the PR, thanks.

But you didn't reply my question. Does it work on your computer now with current code I provided.

I understand you might dislike the current implementation. I agree ot could be improved. But right now,I would like to get a simple reply to the following question.

Does it work?

liveFreeOrCode commented:
> Hi @liveFreeOrCode 
> 
> You commented the PR, thanks.
> 
> But you didn't reply my question. Does it work on your computer now with current code I provided.
> 
> I understand you might dislike the current implementation. I agree it could be improved. But right now,I would like to get a simple reply to the following question.
> 
> Does it work?

Yeah sorry I'm traveling without my PC. I can't test until next week

ccoVeille commented:
OK thanks for replying, we can also wait

LandonTClipp commented:
Closing due to inactivity, feel free to resurrect it if you find the time to finish it.

ccoVeille commented:
Hi @LandonTClipp 

Please reopen.

I was waiting for someone feedbacks.

The code is ready.

sx-klaviyo commented:
I was just searching for `pre-commit` hooks for Mockery and found this 👀  Would love to see this reopened and merged!

ccoVeille commented:
Just ping me back if nothing happen, I can reopen the PR

ccoVeille commented:
@sx-klaviyo 

The MR was waiting for people to test since July. 

@liveFreeOrCode said he would, but he didn't.

There are indeed some code reviews that were provided, but before going further and address them, I would like someone to test.

The code was supposed to work. I would appreciate it if you could, before I iterate on addressing feedbacks.

Because without a tester, it's uneasy to validate things out of my understanding/knowlege

sx-klaviyo commented:
I'll try to test it this week and let you know!

sx-klaviyo commented:
@ccoVeille I confirm that your `.pre-commit-hooks.yaml` is working!
I also left a comment in a the `additional_dependency` and `entry` conversation: https://github.com/vektra/mockery/pull/795#discussion_r1681275093, please let me know if that looks good to you

ccoVeille commented:
Great. Thank you for testing.

For records, I'm unsure I will have time to fix that before Christmas break.

So be patient

sx-klaviyo commented:
Hi @ccoVeille, just wondering if you are back from the Christmas break? Please let me know if I can be help with anything!

ccoVeille commented:
Thanks for the reminder.

Winter break was indeed a great period.

Please let me know, if you think it's OK now.

ccoVeille commented:
In order to test locally, please use the following set up

Add the following lines to your `.pre-commit-config.yaml` file

```yaml
repos:
  - repo: https://github.com/ccoveille-forks/mockery
    rev: v2 # This is a mutable reference to the latest version
    hooks:
      - id: generate-mocks
```

Then test the other ones

```yaml
repos:
  - repo: https://github.com/ccoveille-forks/mockery
    rev: v2 # This is a mutable reference to the latest version
    hooks:
      - id: generate-mocks-local
```

Both should work, I created a fake [2.51.1](https://github.com/ccoveille-forks/vektra-mockery/releases/tag/v2.51.1-alpha) tag

Please check and let me know.

sx-klaviyo commented:
I believe the `repo` link should be `https://github.com/ccoveille-forks/vektra-mockery`
Also, it only worked when I use this:
```yaml
  - repo: https://github.com/ccoveille-forks/vektra-mockery
    rev: v2.51.1-alpha # by explicitly using the tag that you created
    hooks:
      - id: generate-mocks
```
Using `v2` is not working for me. Also, i'm not sure if `generate-mocks-local` is a valid hook: https://github.com/ccoveille-forks/vektra-mockery/blob/0db88da09d747f44f7bdcd2b680e7c77fb4f506e/.pre-commit-hooks.yaml (this is the commit on master branch as I'm making this comment)

ccoVeille commented:
I had to test a lot to finally find why

I worked in a branch in my fork, but here is what the documentation of pre-commit says:

https://pre-commit.com/#updating-hooks-automatically

> You can update your hooks to the latest version automatically by running [pre-commit autoupdate](https://pre-commit.com/#pre-commit-autoupdate). By default, this will bring the hooks to the latest tag on the default branch.

I had to switch the default branch of my fork to be the one of my PR ...

Now everything works as expected. Please check with

```yaml
repos:
  - repo: https://github.com/ccoveille-forks/vektra-mockery
    rev: v2
    hooks:
      - id: generate-mocks
```

Then

```console
$ pre-commit install
$ pre-commit autoupgrade
```

Then try to commit a Go file

Then proceed with the exact same steps with 

```yaml
repos:
  - repo: https://github.com/ccoveille-forks/vektra-mockery
    rev: v2
    hooks:
      - id: generate-mocks-local
```

Then

```console
$ pre-commit install
$ pre-commit autoupgrade
```

And it should be OK (famous last words)

Please let me know @sx-klaviyo and even @liveFreeOrCode 

So much time lost because of the "last tag from the default branch" joke. It's logic, But it's a pain to spot.

I might open a bug report/feature request on pre-commit code

ccoVeille commented:
I'm closing my PR becauit appears to me to be dead end.

- https://github.com/vektra/mockery/issues/791#issuecomment-2645818897

Thank you @liveFreeOrCode @sx-klaviyo for your time, and the tests you made.

Please read my post, you will find the reason of my decision
